### ISSUE TITLE
Console API

### ISSUE THREAD

**Issue by Gustorn** [tomassedovic/tcod-rs#137]
While working on my project I noticed a few things about the `Console` API: the `Root` and `Offscreen` consoles seem to be clearly distinct concepts with some shared common functionality. The current API doesn't really reflect this: all functionality is implemented under the `Console` enum. Examples:

* `init_root` is only applicable to `Root` 
* `new` is only applicable to `Offscreen`
* all static methods could be either free functions or under `RootConsole`. Having them as free functions would probably provide a nicer API, while having them under RootConsole would match the `libtcod` documentation (which is why I'm more in favor of the 2nd option)

My suggestion would be the following: split Console into two structs and a trait:
* `OffscreenConsole`: struct for offscreen console functionality
* `RootConsole`: struct for the root console funcionality
* `Console`: trait for the shared functionality

In practice all trait methods except for `con(&self)` can be implemented in the `Console` trait itself, so `OffscreenConsole` and `RootConsole` would only have to provide `con(&self)`. 

Advantages:
* You can still use `&Console` or `Box<Console>` to have a collection of consoles, if you truly do not want to have distinctions between `Root` and `Offscreen` consoles.
* Type specific functionality will not bleed through to an other type (`new` and `init_root`).

Disadvantages:
* Breaking change
* More import statements (explicit `Console` trait import)

I'm only making these suggestions, because there's an open invitation on the main project page, if you want me to stop, please just say so.

**Comment by tomassedovic** [comment]
Please don't stop, this is the only way we can improve things :-).

(also, sorry, the reply is quite long and I hope it doesn't sound negative, I don't mean it to be, but it's a bit complex and I'd like to have more thinking on this before we go ahead)

The way it's currently organised maps relatively closely to the libtcod interface. Some functions (e.g. `is_fullscreen`, `set_fade`, `flush`) do not require `TCOD_console_t` so I made them as static. Everything that does expect `TCOD_console_t` (e.g. `width`, `put_char`, `clear`) can be applied both to root and offscreen consoles. Apart from the way they're constructed and destroyed, they're pretty much identical (from libtcod's point of view).

This is a bit tough because on one hand, I want good ergonomics and an API that looks closer to Rust than to C, on the other hand I don't want to drastically depart from libtcod because of the people who come here with libtcod knowledge but used a different language so far.

If we made the current static functions into methods on `Root` console, it will limit their use compared to what we have now. You'll have to pass the root console object everywhere you want to read keyboard, etc.

How would we handle blit? A method on the Console trait that takes &mut Console?

I'm not too happy about requiring a new import to even get started, but it's not a deal breaker.

On the other hand, I think it does make sense to me semantically (i.e. that's what I would have done anyway if I were designing libtcod -- which doesn't mean it's necessarily a good idea :-) ).

I'm not too fond of moving the static functions to freestanging ones, because other than the way you type the invocation, they're pretty much identical, aren't they? Thing is -- in that case, I don't see much reason for splitting root and offscreen consoles into different types -- they only thing that'll be different then is the constructor and destructor code. Or am I missing other functions that can only be called on one or the other but not both?

And while we're at this, there's a bit of a weirdness now because we export RootConsole, which you can theoretically write to before `init_root` is called. I don't know what that's going to do, but prolly nothing good. If do make the root console-related functionality into a more "normal" Rust object, we could stop exporting it and only have it available through init_root (or its new equivalent). It would make things slightly more verbose but also safer I think.

Also the fact that getting information is in the system module (check_for_event) but you can get keyboard state from console, too (and it's actually the preferred way as far as I can tell). So should we do something about that? And even more generally, there's an overlap between console and system and it's not always clear why it's where it is (e.g. fullscreen stuff is split, input, flush, fading, fonts, etc.).

**Comment by Gustorn** [comment]
First, now that I though about it, I'm not a big fan free `Console` functions either. Having said that,
you misunderstood the suggestion, probably because it was not clearly written at the first place, so let me go into a bit more detail:

There would be 2 structs and 1 trait, each with the following content:
* RootConsole: struct, contains `init_root` and every static function from the current `Console`, still as static functions (so essentially that change would be, eg.: `Console::flush` into `RootConsole::flush`
* OffscreenConsole: struct, contains `new`
* Console: trait, contains everything from `Console` that takes `&self` or `&mut self`

**Comment by Gustorn** [comment]
I actually made a small-scale implementation of this just to see how it would look, and you were right to mention `blit`, it's the most problematic of the functions. You cannot implement it on the `trait` because it ruins object safety. You could implement it as methods on the concrete `structs`, but that makes it pretty inconvenient to maintain. Here's what I came up with:

Have `blit` as a free function. While I agree, that functions related to the `Root` console should not be free-standing functions, `blit` doesn't really belong to anything. Since `console` is in its separate module (if we count #138) it can be called as `console::blit` and with the same semantics as now.

Then it can be implemented as follows (asserts left out to save space):
```rust
pub fn blit<T, U>(source_console: &T,
                  source_x: i32, source_y: i32,
                  source_width: i32, source_height: i32,
                  destination_console: &mut U,
                  destination_x: i32, destination_y: i32,
                  foreground_alpha: f32, background_alpha: f32)
    where T: Console,
          U: Console {
    unsafe {
        ffi::TCOD_console_blit(source_console.con(),
                               source_x, source_y, source_width, source_height,
                               destination_console.con(), 
                               destination_x, destination_y,
                               foreground_alpha, background_alpha);
    }
}
```

Now if we implement the `Console` trait for various type, using `blit` is actually really convenient. I put together this example file to test the changes:

```rust
extern crate tcod;

use tcod::console;
use tcod::{Console, RootConsole, OffscreenConsole};
use tcod::colors;


fn main() {
    let mut root = RootConsole::init(80, 50, "Using blit with libtcod", false);
    
    let mut direct: OffscreenConsole = OffscreenConsole::new(20, 20);
    let mut boxed_direct: Box<OffscreenConsole> = Box::new(OffscreenConsole::new(20, 20));
    let mut trait_object: &Console = &OffscreenConsole::new(20, 20);
    let mut boxed_trait: Box<Console> = Box::new(OffscreenConsole::new(20, 20));


    root.set_default_background(colors::darkest_green);
    
    direct.set_default_background(colors::red);
    boxed_direct.set_default_background(colors::white);
    trait_object.set_default_background(colors::black);
    boxed_trait.set_default_background(colors::blue);

    root.clear();

    direct.clear();
    boxed_direct.clear();
    trait_object.clear();
    boxed_trait.clear();

    
    console::blit(&direct, 0, 0, 20, 20, &mut root, 0, 0, 1.0, 1.0); 
    console::blit(&boxed_direct, 0, 0, 20, 20, &mut root, 20, 0, 1.0, 1.0); 
    console::blit(&trait_object, 0, 0, 20, 20, &mut root, 0, 20, 1.0, 1.0); 
    console::blit(&boxed_trait, 0, 0, 20, 20, &mut root, 20, 20, 1.0, 1.0); 

    RootConsole::flush();
    RootConsole::wait_for_keypress(true);
}

```

**Comment by Gustorn** [comment]
I wrote up some comments previously, but after much delibaration I decided to rewrite them into a more appealing format. My earlier posts contained an explanation detailing my original suggestion, but right now, I don't think that suggestion went far enough. In the next several points I'll list the issues I see with the current design, the propsed solutions then some cons for the whole change. Fair warning: this is going to be a really long post.

Object Initialization 
-------------------------
**The problem:** the problem consists of three parts:
1. `RootConsole` can be written to a read from before it is initialized, with methods that take `&self` or `&mut self`
2. some of its static methods can only be called **before** `init`
3. some of its static methods can only be be called **after** `init`

**The solution:**
1. The proposed change already solves this problem, since `RootConsole::init` is the only way to get an instance of the `RootConsole` object.
2. Since I could only find a single function that requires this (`TCODConsole::setCustomFont`), adding a second `init` function (let's say `RootConsole::init_with_font`) is a feasible solution. An other alternative is building a separate `Settings` object and passing that to `RootConsole`. I'm in favor of the first alternative.
3. This is the trickiest of the three problems, and would require us to convert all of `RootConsole`'s static functions into methods, taking `&self` or `&mut self`. I firmly believe this is the right move, but this leads up to the second issue:

Thread safety
-------------------
**The problem:** as of right now, we are not treating the `RootConsole` as an object. We are wrapping functions that essentially change mutable global state (or `static mut` in Rust terms), while making no guarantees about the safety of these operations. Consider the following code:

```rust
fn main() {
    Console::init_root(80, 20, "Threading bug", false);

    for i in 0..2 {
        thread::spawn(move || {
            Console::set_fullscreen(Console::is_fullscreen());
        }); 
    }

    timer::sleep(Duration::milliseconds(50));
}
```

Since the borrow checker doesn't know we are manipulating some global state from different threads, this compiles without even a warning. This code actually **segfaults** on my system, which is definitely a bug.

**The solution:** same as the previous issue: declaring the `RootConsole` static functions as actual methods would essentially make `RootConsole` into something the borrow checker can validate. I have a hard time thinking about alternative solutions (locking on a `static` would probably work, but it's not very elegant).

Cons
-------
Divergence from the original `C` API: while this is a valid concern, I think it's more important to provide something that reflects Rust's philosophy. The functionality is still very recognizable, but it's wrapped in a way that makes sense in the Rust ecosystem.

Also there's the concern that changing the previously static functions into methods will limit the functionality. While this is true, as I detailed in my `Thread safety` section, the current implementation is inherently unsafe. I also think that this change would be a nice step towards a nice Rust-like API.

Implementation Details
-------------------------------
All the changes detailed above are fairly easy to do. I did a proof of concept implementation to see, how it'd look, and the only problem I ran into was the `blit` function (very correctly pointed out in your previous comment). 

**The problem** is that it cannot be implemented on the `Console` trait, since it would ruin object safety, which would in turn make creating `Box<Console>` objects impossible. This, of course had to be avoided. An other alternative would've been to implement it for all the concrete `Console` implementers (`RootConsole` and `OffscreenConsole`). This would unfortunately lose out on generality: traits objects would no longer be able to use `blit`. This was also not an option.

**The solution** was to have `blit` as a free function. Now you said, that you disliked the idea of free functions here, but hear me out: `blit` doesn't logically belong anywhere, even in the current implementation. It is a function, that operates on any two `Console` types, but is not a method. If/when #138 gets merged, `console` will be split into its own module, which made `blit` a nice candidate for a free function, since the calling syntax would be the following:

```rust
use tcod::console;

console::blit(...)
```
Which is really close to the current implementation, but with all the added benefits of the points mentioned in the post. It's also worth noting, that we wouldn't lose out on performance either: with the right `Console` trait implementation the following `blit` implementation is possible (asserts ommited to save space):

```rust
pub fn blit<T, U>(source_console: &T,
                  source_x: i32, source_y: i32,
                  source_width: i32, source_height: i32,
                  destination_console: &mut U,
                  destination_x: i32, destination_y: i32,
                  foreground_alpha: f32, background_alpha: f32)
    where T: Console,
          U: Console {
    unsafe {
        ffi::TCOD_console_blit(source_console.con(),
                               source_x, source_y, source_width, source_height,
                               destination_console.con(), 
                               destination_x, destination_y,
                               foreground_alpha, background_alpha);
    }
}
```

It's also worth mentioning, that this would also make it far more convenient to use for boxed and trait objects. This is from the example file I wrote to test the reworked `blit` function:

```rust
let mut root = RootConsole::init(80, 50, "Using blit with libtcod", false);

let mut direct: OffscreenConsole = OffscreenConsole::new(20, 20);
let mut boxed_direct: Box<OffscreenConsole> = Box::new(OffscreenConsole::new(20, 20));
let mut trait_object: &Console = &OffscreenConsole::new(20, 20);
let mut boxed_trait: Box<Console> = Box::new(OffscreenConsole::new(20, 20));


console::blit(&direct, 0, 0, 20, 20, &mut root, 0, 0, 1.0, 1.0); 
console::blit(&boxed_direct, 0, 0, 20, 20, &mut root, 20, 0, 1.0, 1.0); 
console::blit(&trait_object, 0, 0, 20, 20, &mut root, 0, 20, 1.0, 1.0); 
console::blit(&boxed_trait, 0, 0, 20, 20, &mut root, 20, 20, 1.0, 1.0); 
```
Note that you don't have to stack `&`s and `*`s to get `blit` working.

Edits
-------

**Comment by tomassedovic** [comment]
Sorry I misread your original idea even though the new proposal is about making these functions members of RootConsole.

Secondly, while I do want to keep tcod-rs familiar to existing libtcod users, if a user gets a segfault while using the "safe" parts of tcod, that's absolutely a bug.

And yeah, I think your suggestion makes sense. Methods on root console will let the borrow checker deal with the shared mutable state that libtcod tracks and I'm in favour of your blit suggestion, too.

Regarding "init root with font": I think you can use it before _and_ after the root console is initialised. If you call it after a root console is initialised, it'll just change the way things are rendered. So I think we may want to keep both "init with font" constructor and keep the "set_custom_font" as a method around, too.

Also, I would actually like to require picking a font when creating a root console as the default interface. `terminal.png` is just a fallback when you don't enter anything and it's kept for legacy reasons. It's super ugly and doesn't seem to follow the layout of the newer fonts shipped with tcod.

Problem is, it's another four arguments to the constructor.

**Comment by Gustorn** [comment]
Well, if you can set the font before and after, then `libtcod`'s [documentation](http://roguecentral.org/doryen/data/libtcod/doc/1.5.1/html2/console_set_custom_font.html?c=true&cpp=false&cs=false&py=false&lua=false) is pretty broken :-). 

I actually thought about the problem of having a lot of parameters (both to functions and constructors), and I can't really see an easy fix. We can keep it as it is, and rely on the user to check the docs. An other idea might be to have separate objects that we pass to these functions, as to imitate named parameters. I'm not sure I'm too fond of this idea either.

**Comment by tomassedovic** [comment]
Hm, I think I may have misinterpreted the docs. Anyway, I've tried it now and it doesn't seem to have an effect, so I'm fine with just allowing that during initialisation.

Regarding parameters, not sure if this is what you meant by the second idea, but something like [http://doc.rust-lang.org/std/process/struct.Command.html](http://doc.rust-lang.org/std/process/struct.Command.html) doesn't seem too bad to me:

```rust
let mut root = tcod::console::Root::new()
                      .size(80, 40),
                      .fullscreen()
                      .font(&Path("dejavu16x16_gs_tc.png").
                      .initialize();

root.print_ex(40, 15, BackgroundFlag::None, TextAlignment::Center,"Hello!");
root.flush();
```

**Comment by tomassedovic** [comment]
Although since `Root::new` doesn't actually return an instance of `Root` but some intermediate builder struct that actually has the `size`, `font`, etc. methods, this would be both surprising and split things in the docs. So maybe not really worth it.

**Comment by Gustorn** [comment]
My suggestion was making a separate `RootData` or something like that, which would make the initialization into something like this:

```rust
let mut root = tcod::console::Root::init(RootData { width: 80, height: 30,
                                                    is_fullscreen: true,
                                                    font: &Path("dejavu16x16_gs_tc.png" });
                                                            
```

This is of cource just a poor man's replacement for named parameters, but apart from this and the builder pattern you mentioned I don't know what other possibilities there are. Maybe a different naming convention for the pattern you mentioned?

```rust
let mut root = tcod::console::Root::initializer()
                      .size(80, 40),
                      .fullscreen()
                      .font(&Path("dejavu16x16_gs_tc.png").
                      .create();

root.print_ex(40, 15, BackgroundFlag::None, TextAlignment::Center,"Hello!");
root.flush();
```

This has the problem that the user might think that they can create `Root` consoles at will, so this is not a terribly good idea either.

On a slightly different topic: would you prefer dropping `Console` suffixes in favor of `console::Root` and no reexports?

**Comment by Gustorn** [comment]
Alternatively, we could use a slightly modified version of your idea:
```rust
use tcod::console as console;

let mut root = console::RootInitializer::new()
                      .size(80, 40),
                      .fullscreen()
                      .font(&Path("dejavu16x16_gs_tc.png").
                      .initialize();
```

This has the advantage that the documentation would not have to be split, `Root` would still have its `init` static function taking a lot of parameters and `RootInitializer` would just be a convenience struct, making its initialization easier.

**Comment by tomassedovic** [comment]
I like your last suggestion.

Also, I think we can combat the lack of explicitness in the full `init_root` by using better values for the parameters. Say instead of passing `bool` to fullscreen, we could have an enum that is `Windowed` or `Fullscreen`. The screen dimensions, title and font path are relatively straightforward and we could create a named tuple for the font char layout:

```rust
let mut root = console::Root::init(80, 40, "My Game", Fullscreen,
    &Path("dejavu16x16_gs_tc.png"), FONT_LAYOUT_TCOD,
    FontCharacterDimension(32, 8));
```

Still not perfect but a bit more readable imho.

**Comment by Gustorn** [comment]
While we're at naming, we could actually create a simple additional `Font` type and make the `Path` inside of the `Console`, which would make the above example:
```rust
let mut root = console::Root::init(80, 40, "My Game", Fullscreen,
    Font { path: "dejavu16x16_gs_tc.png" } , FONT_LAYOUT_TCOD,
    FontCharacterDimension(32, 8));

// Or 
let mut root = console::Root::init(80, 40, "My Game", Fullscreen,
    Font::with_path("dejavu16x16_gs_tc.png") , FONT_LAYOUT_TCOD,
    FontCharacterDimension(32, 8));

// Or 
let mut root = console::Root::init(80, 40, "My Game", Fullscreen,
    Font("dejavu16x16_gs_tc.png"), FONT_LAYOUT_TCOD,
    FontCharacterDimension(32, 8));
```

**Comment by Gustorn** [comment]
Since we still seem to be discussing this.

**Comment by tomassedovic** [comment]
Closed via #144.
